examples/knative/go/kn-go-tagging/tagging/tagging_test.go

```diff
@@ -0,0 +1,396 @@
+package tagging
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/cloudevents/sdk-go/v2/event"
+	"github.com/vmware/govmomi/simulator"
+	_ "github.com/vmware/govmomi/vapi/simulator"
+	"github.com/vmware/govmomi/vapi/tags"
+	"github.com/vmware/govmomi/vim25"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+	"go.uber.org/zap/zaptest"
+	"knative.dev/pkg/logging"
+)
+
+func TestParseEventMoRef(t *testing.T) {
+	type vm struct {
+		vmType string
+		Value  string
+	}
+
+	var tests = []struct {
+		name      string
+		path      string
+		expectErr bool
+		want      *vm
+	}{
+		{
+			"Test that event is readable",
+			"testdata/event.json",
+			false,
+			&vm{vmType: "VirtualMachine", Value: "vm-10000"},
+		},
+		{
+			"Event should return error if vm type and value are null",
+			"testdata/eventErr1.json",
+			true,
+			nil,
+		},
+		{
+			"Event should return error if vm info is null",
+			"testdata/eventErr2.json",
+			true,
+			nil,
+		},
+		{
+			"Event should return error if vm parent is null",
+			"testdata/eventErr3.json",
+			true,
+			nil,
+		},
+		{
+			"Event should return error if data is null",
+			"testdata/eventErr4.json",
+			true,
+			nil,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			e := jsonFileToEvent(tt.path)
+			moRef, err := parseEventMoRef(e)
+
+			if tt.expectErr && err == nil {
+				t.Error("expected error, but got none")
+			}
+
+			if !tt.expectErr && err != nil {
+				if moRef.Type != tt.want.vmType {
+					t.Errorf("got: %q, want: %q", moRef.Type, tt.want.vmType)
+				}
+
+				if moRef.Value != tt.want.Value {
+					t.Errorf("got: %q, want: %q", moRef.Value, tt.want.Value)
+				}
+			}
+		})
+	}
+}
+
+func TestRun(t *testing.T) {
+	const (
+		username = "administrator@vsphere.local"
+		password = "pass"
+	)
+
+	secretsDir := createSecret(t, username, password)
+	defaultEnv := EnvConfig{
+		Port:       50001,
+		SecretPath: secretsDir,
+		Insecure:   true, // vcsim
+	}
+
+	t.Run("fail with authentication error", func(t *testing.T) {
+		model := simulator.VPX()
```

Comment:
hint (non-blocking): if these sub-tests have no dependencies/are thread-safe, you could add `t.Parallel()` to speed up tests further. in this case you would have to move `defaultEnv` to each sub-tests to avoid races though